Fast Pair BLE Device Discovery for Windows in Rust

TheShepord opened the pull request:
<!--
  Thanks for submitting a pull request!
  We appreciate you spending the time to work on these changes. Please provide
  enough information so that others can review your pull request.
-->

## Summary
Implementing Fast Pair BLE Device Discovery for Windows in Rust, as part of intern project. See go/fastpair-seeker-rust for further discussion on design decisions (Design > API Decisions).

## How did you test this change?
This is an in-development prototype, so unit and integration tests will be provided at a later date (b/288592509). 
`cargo run` output for Windows: 
![image (1)](https://github.com/google/nearby/assets/53844396/93872fe6-664e-4866-90e6-e470f225b702)

`cargo run` output for unsupported platform:

![image (2)](https://github.com/google/nearby/assets/53844396/e014f87a-c5de-48d3-b327-8161de0093b1)

anayw2001 commented on fastpair/rust/src/bluetooth/mod.rs:
```diff
@@ -0,0 +1,43 @@
+        mod windows_ble;
+        use windows_ble as imp;
+    } else {
+        compile_error!("Unsupported target platform. Currently supported: Windows.");
```
may be better to just panic here instead of compile_error just so github actions can pass. would you also mind adding a build rust on windows github action to test the new windows code?

anayw2001 commented on fastpair/rust/src/bluetooth/windows_ble/adapter.rs:
```diff
@@ -35,6 +73,87 @@ impl BleAdapter {
+
+    pub fn scanner(&self) -> Result<impl Stream<Item = BleDevice>, anyhow::Error> {
+        let watcher = BluetoothLEAdvertisementWatcher::new()?;
+        watcher.SetScanningMode(BluetoothLEScanningMode::Active)?;
```
this didn't work for me afaik, so maybe catching an error here and logging a warning instead may be a better approach.

TheShepord replied:
Unsure what you mean by this, which part didn't work for you and which warning should I be logging?

anayw2001 replied:
the SetScanningMode to be specific, apparently it's a hardware-dependent call? weird but yeah definitely log a warning if that was unable to be set, but i should imagine the applet would still work.

anayw2001 commented on fastpair/rust/src/bluetooth/windows_ble/adapter.rs:
```diff
@@ -35,6 +73,87 @@ impl BleAdapter {
+        }
+
+        let (sender, receiver) = futures::channel::mpsc::channel(16);
+        let sender = Arc::new(std::sync::Mutex::new(sender));
```
why is this needed?

anayw2001 replied:
nvm

TheShepord replied:
Added comment to prevent further confusion.

TheShepord commented:
> may be better to just panic here instead of compile_error just so github actions can pass. would you also mind adding a build rust on windows github action to test the new windows code?

Added `unsupported` module, calling any methods cause a panic.


> this didn't work for me afaik, so maybe catching an error here and logging a warning instead may be a better approach. 

Unsure what you mean by this, could you elaborate?

> why is this needed?

Added comments to prevent further confusion.

TheShepord commented:
Changed from wrapper to trait-based implementation for cross-platform support. General cleanup. @anayw2001, @suetfei.

marshallpierce commented on fastpair/rust/rustfmt.toml:
```diff
@@ -0,0 +1 @@
+max_width = 80
```
why so short?

anayw2001 replied:
g3 rustfmt will reformat these if saved in cider or something so that's probably why.

marshallpierce commented on fastpair/rust/src/bluetooth/windows_ble/adapter.rs:
```diff
@@ -0,0 +1,177 @@
+            // Move `event_args` into async block.
+            async move {
+                match event_args.AdvertisementType().ok()? {
+                    BluetoothLEAdvertisementType::NonConnectableUndirected => None,
```
Why are these skipped?

anayw2001 replied:
i suspect fastpair doesn't need these, but i will probably go through this at some point later and make an enum for the types to be exported back to the caller.

marshallpierce commented:
Also, I suggest running `rustfmt` and `cargo clippy --all-targets` to get hints on stuff to clean up

anayw2001 commented on fastpair/rust/src/bluetooth/common.rs:
```diff
@@ -0,0 +1,47 @@
+    // should only ever be one concrete type implementing Adapter. Change this
+    // to `impl Stream` once impl trait return types are stabilized in traits.
+    // b/289224233.
+    fn scan_devices(&self) -> Result<Pin<Box<dyn Stream<Item = Self::TDevice>>>, anyhow::Error>;
```
could we do something like next_device(&self) and have it block until a device is seen? that way we can put it in a nice loop {} block.

TheShepord replied:
I'm fairly certain this is already the case? Stream provides poll_next, so we can do stream.next().await as shown in main.rs. Is this not what you mean?

anayw2001 replied:
just in a slight effort to make this return type slightly less ugly tbh

anayw2001 commented on fastpair/rust/src/bluetooth/common.rs:
```diff
@@ -0,0 +1,47 @@
+// such as pairing.
+pub trait Device {
+    // Retrieve the name advertised by this device.
+    fn name(&self) -> Result<String, anyhow::Error>;
```
maybe getting mac address?

TheShepord replied:
I'll add additional calls as needed, there's a lot of information that could be nice to fetch.

anayw2001 commented on fastpair/rust/src/bluetooth/unsupported/adapter.rs:
```diff
@@ -0,0 +1,49 @@
+
+// Concrete type implementing `Adapter`, used for unsupported devices.
+// Every method should panic.
+pub struct BleAdapter {}
```
can just be pub struct BleAdapter;

suetfei commented on fastpair/rust/src/lib.rs:
```diff
@@ -0,0 +1,17 @@
+
+// Library file, exports modules for use in integration tests and external crates.
+
+pub mod bluetooth;
```
They are not used now, but will be later for Fastpair V2 or whatever.

TheShepord commented on fastpair/rust/src/main.rs:
```diff
@@ -12,9 +12,22 @@
+use futures::executor;
+use futures::stream::StreamExt;
+
+fn main() -> Result<(), anyhow::Error> {
```
For now I'll leave it here since I'm developing a scanner demo as my intern project. The most likely restructure afterwards is making bluetooth a standalone top-level crate alongside a separate scanner crate. Thank you!

TheShepord commented on fastpair/rust/src/bluetooth/windows_ble/device.rs:
```diff
@@ -0,0 +1,50 @@
+    ) -> Result<Self, anyhow::Error> {
+        let inner =
+            BluetoothLEDevice::FromBluetoothAddressWithBluetoothAddressTypeAsync(addr, kind)?
+                .await?;
```
I'm against making this a tuple struct to facilitate adding more parameters in the future. Map here doesn't quite work unfortunately, error types are mismatched. I'll fix up the error types as per your other comment, let me know if the new design works.

TheShepord commented on fastpair/rust/src/bluetooth/windows_ble/device.rs:
```diff
@@ -0,0 +1,50 @@
+
+impl BleDevice {
+    pub async fn from_addr(
+        addr: u64,
```
I'll leave as-is for now since this is a private method and Windows bluetooth addresses use u64. Will make a note for future.

TheShepord commented on fastpair/rust/src/bluetooth/mod.rs:
```diff
@@ -0,0 +1,25 @@
+cfg_if::cfg_if! {
+    if #[cfg(windows)] {
+        mod windows_ble;
+        pub use windows_ble::*;
```
To be clear, are you suggesting a top-level wrapper struct that calls platform-specific structs? Or rather, are you suggesting explicitly exporting every struct (e.g. pub use windows_ble::BleAdapter as BleAdapter)?

TheShepord commented on fastpair/rust/Cargo.toml:
```diff
@@ -20,3 +20,15 @@ edition = "2021"
 # See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
 
 [dependencies]
+anyhow = "1.0"
```
I'll file a bug and fix the error types once the bluetooth module is split into its own crate.

TheShepord commented on fastpair/rust/src/bluetooth/common.rs:
```diff
@@ -0,0 +1,47 @@
+// Concrete types implementing this trait are Bluetooth Central devices.
+// They provide methods for retrieving nearby connections and device info.
+pub trait Adapter {
+    type TDevice: Device;
```
I've seen this employed in libraries as a shorthand for "Type." Happy to change if this isn't actually a thing.

marshallpierce replied:
I don't see it in practice. Maybe some C# refugee developer who still bears the scars of previous suffering. Nowadays anyone who's using an IDE that isn't syntax highlighting associated types is just a masochist.

TheShepord commented on fastpair/rust/src/bluetooth/common.rs:
```diff
@@ -0,0 +1,47 @@
+    type TDevice: Device;
+
+    // Retrieve the system-default Bluetooth adapter.
+    async fn default() -> Result<Box<Self>, anyhow::Error>;
```
Filed a bug, I'll get around to this once my project demo is finalized.

TheShepord commented on fastpair/rust/src/bluetooth/unsupported/mod.rs:
```diff
@@ -0,0 +1,21 @@
+
+// Bluetooth LE module for unsupported devices. Every method panics.
+
+mod adapter;
```
To be clear, you're suggesting I make the unsupported module a single unsupported.rs file as opposed to a folder?

marshallpierce replied:
Yep, that'd work.

anayw2001 reviewed (approved):
lgtm! stream type is annoying, but we'll figure it out _eventually_.

marshallpierce commented on fastpair/rust/src/bluetooth/unsupported/adapter.rs:
```diff
@@ -0,0 +1,49 @@
+    type Device = BleDevice;
+
+    async fn default() -> Result<Self, anyhow::Error> {
+        panic!("Unsupported target platform.");
```
there is also `unimplemented!` which does the same thing, and maybe is more communicative to the reader? Or not. I'm not sure which is better.

TheShepord replied:
I personally prefer `panic!` here. If there's a specific platform with pending features then `unimplemented!` makes more sense, but for an entire unsupported platform what's unimplemented isn't the specific function, but rather the entire platform library.
